cmd/process-agent/collector.go

```diff
@@ -79,13 +79,37 @@ type Collector struct {
 	runRealTime bool
 }
 
+// initProcessChecksRuntime initializes global variables that are shared across pkg/process/checks
+func initProcessChecksRuntime() {
+	batchSize := ddconfig.Datadog.GetInt("process_config.max_per_message")
+	if batchSize <= 0 {
+		log.Warnf("Invalid item count per message (<= 0), using default value of %d", ddconfig.DefaultProcessMaxPerMessage)
+		batchSize = ddconfig.DefaultProcessMaxPerMessage
+	} else if batchSize > ddconfig.DefaultProcessMaxPerMessage {
+		log.Warnf("Overriding the configured max of item count per message because it exceeds maximum limit of %d", ddconfig.DefaultProcessMaxPerMessage)
+		batchSize = ddconfig.DefaultProcessMaxPerMessage
+	}
+	checks.MaxBatchSize = batchSize
+
+	ctrProcsBatchSize := ddconfig.Datadog.GetInt("process_config.max_ctr_procs_per_message")
+	if ctrProcsBatchSize <= 0 {
+		log.Warnf("Invalid max container processes count per message (<= 0), using default value of %d", ddconfig.DefaultProcessMaxCtrProcsPerMessage)
+		ctrProcsBatchSize = ddconfig.DefaultProcessMaxCtrProcsPerMessage
+	} else if ctrProcsBatchSize > ddconfig.ProcessMaxCtrProcsPerMessageLimit {
+		log.Warnf("Overriding the configured max of container processes count per message because it exceeds maximum limit of %d", ddconfig.ProcessMaxCtrProcsPerMessageLimit)
+		ctrProcsBatchSize = ddconfig.DefaultProcessMaxCtrProcsPerMessage
+	}
+	checks.MaxCtrProcsBatchSize = ctrProcsBatchSize
+}
+
 // NewCollector creates a new Collector
 func NewCollector(cfg *config.AgentConfig) (Collector, error) {
 	sysInfo, err := checks.CollectSystemInfo(cfg)
 	if err != nil {
 		return Collector{}, err
 	}
 
+	initProcessChecksRuntime()
```

Comment:
What about just creating a `collectorConfig` struct that would house all collector specific configs. So we create the initialize that config here. It would require modification to the check interface to support passing it in though, but maybe we replace the broader `AgentConfig` with a smaller one which we control?

Reply:
The idea of the refactor is to move settings away from the `AgentConfig` and instead read them directly from the global `Config` object.

The batch size is a check property so ideally we should initialize it within each one of them.

The main problem here is that we can't pass check specific configs using the current `Check` interface and we have a singleton for each one of the checks that we set up with the `Init` method.
https://github.com/DataDog/datadog-agent/blob/b5bed4dc6a5b08a50346e614b386ea67e37695de/pkg/process/checks/checks.go#L13-L22

e.g. `Process` check
https://github.com/DataDog/datadog-agent/blob/b5bed4dc6a5b08a50346e614b386ea67e37695de/pkg/process/checks/process.go#L28-L29

Ideally what we should have is a constructor for each check that takes specific configs. Instead of using `Init` to initialize the check properties, we could have these constructors instantiating the singletons with specific configs read from the global `Config` object.

Reply:
If the idea of sentinel exported vars isn't getting appreciated, I think the next best options are:
- Following @wiyu's suggestion - to add the `CollectorConfig` struct that really contains simple init values and replace `AgentConfig` in the interface with that
- Implement `New<CheckType>` initializers with these params

Also we probably should eventually remove any kind of configuration from `Run` and `RunWithOptions`.

Reply:
I've ended up moving the batch size initialization to the `pkg/checks` package using two auxiliary functions: `getMaxBatchSize` and `getMaxCtrProcsBatchSize`. This way we read/validate the settings closer to the objects that are actually using them.

Each check calls them during initialization and I added a `sync.Once` to avoid spamming the same log message in case the user sets an invalid setting. What do you think about this approach? 

 To follow up on the suggestions:
> to add the `CollectorConfig` struct that really contains simple init values and replace `AgentConfig` in the interface with that

Our goal is to only keep relevant settings in the `AgentConfig` struct. In this case, wouldn't we be just replacing `AgentConfig` with another name? Correct me if I'm wrong but isn't it the same as keeping `MaxPerMessage` and `MaxCtrProcessesPerMessage` in `AgentConfig`?

> Implement `New<CheckType>` initializers with these params

To do so, we need to get rid of the singleton pattern used by all process-agent checks. I think that this is actually a better solution and something worth doing. However, I'd prefer to tackle it separately since it affects all checks initialization and seems a change involving more risks.